stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/experimental/ChatSocket.kt

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright (c) 2014-2022 Stream.io Inc. All rights reserved.
+ *
+ * Licensed under the Stream License;
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://github.com/GetStream/stream-chat-android/blob/main/LICENSE
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.getstream.chat.android.client.socket.experimental
+
+import io.getstream.chat.android.client.LifecycleHandler
+import io.getstream.chat.android.client.StreamLifecycleObserver
+import io.getstream.chat.android.client.clientstate.DisconnectCause
+import io.getstream.chat.android.client.errors.ChatError
+import io.getstream.chat.android.client.errors.ChatErrorCode
+import io.getstream.chat.android.client.errors.ChatNetworkError
+import io.getstream.chat.android.client.events.ChatEvent
+import io.getstream.chat.android.client.events.ConnectedEvent
+import io.getstream.chat.android.client.events.HealthEvent
+import io.getstream.chat.android.client.models.User
+import io.getstream.chat.android.client.network.NetworkStateProvider
+import io.getstream.chat.android.client.socket.HealthMonitor
+import io.getstream.chat.android.client.socket.SocketFactory
+import io.getstream.chat.android.client.socket.SocketListener
+import io.getstream.chat.android.client.socket.experimental.ChatSocketStateService.State
+import io.getstream.chat.android.client.socket.experimental.ws.StreamWebSocket
+import io.getstream.chat.android.client.socket.experimental.ws.StreamWebSocketEvent
+import io.getstream.chat.android.client.token.TokenManager
+import io.getstream.chat.android.client.utils.stringify
+import io.getstream.chat.android.core.internal.coroutines.DispatcherProvider
+import io.getstream.logging.StreamLog
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.launch
+
+@Suppress("TooManyFunctions", "LongParameterList")
+internal class ChatSocket private constructor(
+    private val apiKey: String,
+    private val wssUrl: String,
+    private val tokenManager: TokenManager,
+    private val socketFactory: SocketFactory,
+    private val coroutineScope: CoroutineScope,
+    private val lifecycleObserver: StreamLifecycleObserver,
+    private val networkStateProvider: NetworkStateProvider,
+) {
+    private var streamWebSocket: StreamWebSocket? = null
+    private val logger = StreamLog.getLogger("Chat:Socket")
```

Comment:
Would you mind marking this particular `Socket` implementation with a slightly different tag? 
`"Chat:Socket-New"` maybe?